### ISSUE TITLE
[feat] Render winit subwindow within Tauri app

### ISSUE THREAD

**Issue by tqwewe** [tauri-apps/tauri#3662]
### Describe the problem

I'd like to render bevy app within Tauri natively. Bevi uses winit to render apps.

### Describe the solution you'd like

Have a child winit app within my main window which can be positioned with a fixed padding of the main window.

### Alternatives considered

- Wasm canvas, though it turned out to be quite slow.
- Create the separate bevy window, set it to borderless and manually move it with the tuari window.

### Additional context

My main goal is to make a bevy editor with tauri, where the middle "viewport" is made in bevy, and the surrounding UI is tauri html and css elements. Similar to the structure of Unity.

**Comment by amrbashir** [comment]
you can use `app.create_tao_window(). 

tao is just a fork of winit but it won't work on Linux though because, tao uses a different backend on Linux.

**Comment by amrbashir** [comment]
I don't think this is possible currently for two reasons:
1. we use `tao` which is a fork of `winit` but with a different Linux backend, so `bevy` won't work on Linux with Tauri.
2. To achieve what you need, you probably need a way to lay the webview on top of the `tao` window, and have it pass clicks to `bevy` and you probably need to control the webview transparency apart from the window itself. Or you need to have multiple webviews which you can position and resize inside the window so you don't have to pass any clicks. Either way needs you to use `wry` directly and also needs modifications at `wry` too to allow resizing and positioning a webview inside a window.

**Comment by tqwewe** [comment]
@amrbashir thanks for the fast response. I've just managed to clone bevy locally and switch winit to tao quite easily with a few small code reference changes.

Now that I have bevy running in tao, could you point me in the right direction to having my bevy app running within the main tauri window?

I'd like to have bevy app placed in the center of my tauri app with empty space surrounding it for my html UI to be placed.

**Comment by amrbashir** [comment]
> @amrbashir thanks for the fast response. I've just managed to clone bevy locally and switch winit to tao quite easily with a few small code reference changes.

I suspect you're on Windows or macOS because this shouldn't work on Linux at all, because bevy probably can't get a gl surface on Linux using tao. 

> I'd like to have bevy app placed in the center of my tauri app with empty space surrounding it for my html UI to be placed.

Yeah, in order for this to happen you need to modify `wry` to expose apis to allow resizing the webview to a certain size and allow positioning it inside the window because currently a webview will just fill the whole window surface.

After that, you need to expose the same `wry` apis on `tauri` too. 

But you need to confirm that bevy works with tao on Linux first so you don't waste your time.

**Comment by tqwewe** [comment]
Damn thank you for the information.
The main benefit of using tao with bevy is that I can make the UI with HTML & CSS (really customizable) and still have the native bevy viewport.

Are there any alternatives I could use? I'm familiar with iced ui library, but I think html and css are really powerful for this kind of thing.

**Comment by FabianLars** [comment]
> Are there any alternatives I could use?

What's wrong with bevy's UI? Sure it's not as flexible as html,css,js but at least you're not adding a performance hog to your game(?).
Ofc this doesn't matter for non-fps heavy games, so feel free to ignore my comment 👍

**Comment by tqwewe** [comment]
> What's wrong with bevy's UI? Sure it's not as flexible as html,css,js but at least you're not adding a performance hog to your game(?).

I'm actually looking into making a bevy editor, and exploring my options. I haven't really worked with bevy's in-built UI, I don't know how flexible it is.

**Comment by FabianLars** [comment]
Ah sorry, overlooked the editor comment 😅 

Edit: But depending on what people create with your editor, slow web tech could indeed end up being problematic 🤔

**Comment by tqwewe** [comment]
I was thinking the web part wouldn't have to be integrated with bevy, it just has to communicate it. So there would be a gap in the middle where bevy would be rendered, and the sorrounding area is just normal tauri UI, which shouldn't have any concern with performance.

But as stated, the lack of linux support is quite frustrating, so I might have to just go with egui or something :/

**Comment by tqwewe** [comment]
@amrbashir I'm going down the path of trying to modify wry to manage the transparent pass manually, and attempting to resize the webview manually too.

> you probably need to control the webview transparency apart from the window itself

Any pointers on where the transparency should be modified? I tried adding `.with_transparent(true)` to the window, which does make the window transparent, but the bevy render is not visible below.

> Or you need to have multiple webviews which you can position and resize inside the window so you don't have to pass any clicks. Either way needs you to use wry directly and also needs modifications at wry too to allow resizing and positioning a webview inside a window.

I've resized the webview by cloning wry, but bevy doesn't seem to be rendered under the webview. Not sure why, if maybe Tuari is running a clear pass which is removing what bevy renders?

**Comment by amrbashir** [comment]
You should try using `wry` directly without `tauri`.

> Any pointers on where the transparency should be modified?

With `wry` you can already control transparency of the webview apart from the window itself. Tauri combines both window and webview transparency into one option.

> I've resized the webview by cloning wry, but bevy doesn't seem to be rendered under the webview. Not sure why, if maybe Tuari is running a clear pass which is removing what bevy renders?

Interesting, which OS?

**Comment by tqwewe** [comment]
> You should try using wry directly without tauri.

Oh okay, I didnt of going directly with wry like that.

> With wry you can already control transparency of the webview apart from the window itself. Tauri combines both window and webview transparency into one option.

Yeah I managed to make the webview transparent, but couldn't see the bevy render behind. I thougnt maybe the transparency pass was clearning the bevy render.

> Interesting, which OS?

I'm on MacOS (Intel). I'm not sure why it wouldn't be rendering bevy... I was thinking maybe tao window lacked support for it, but I was able to modify bevy_winit to use tao and the examples still work... so I'm not sure now.

Here's the repo I'm working with:
https://github.com/tqwewe/bevy-editor-2

You can see I set the webview frame size here:
https://github.com/tqwewe/bevy-editor-2/blob/main/src-tauri/wry/src/webview/wkwebview/mod.rs#L399

**Comment by amrbashir** [comment]
> I'm on MacOS (Intel).

I suspected that. I think the webview replaces the `NSView` of the `tao` window and that removes bevy rendering. I am not expert in macOS development so I can't really help there but I am 90% sure it would work on Windows.

**Comment by tqwewe** [comment]
Ohh interesting, I'll try to ask someone on windows to test it.

**Comment by JoeHowarth** [comment]
Hey @tqwewe did you ever get this working? I'm interested in something similar

**Comment by tqwewe** [comment]
> Hey @tqwewe did you ever get this working? I'm interested in something similar

Sadly nope :') I have since given up on this whole thing. I'm sure some people much smarter than me will end up making an incredible bevy editor 🥹

**Comment by amrbashir** [comment]
A bit of good news, we recently refactored `wry` (which is the underlaying WebView library used by tauri) to make it work with `winit` and other window libraries that has a raw-window-handle, you can even have multiple webviews in the same window, and you can mix it with a wgpu or opengl rendering app, and with bevy as well. However, you'll be limited to macOS, Windows and Linux (X11). Wayland is not supported. Here is an example of wgpu https://github.com/tauri-apps/wry/blob/4f42acb4bd5ea7f16bf0ed18ea8113a9bd957659/examples/wgpu.rs

**Comment by igagen** [comment]
> A bit of good news, we recently refactored `wry` (which is the underlaying WebView library used by tauri) to make it work with `winit` and other window libraries that has a raw-window-handle, you can even have multiple webviews in the same window, and you can mix it with a wgpu or opengl rendering app, and with bevy as well. However, you'll be limited to macOS, Windows and Linux (X11). Wayland is not supported. Here is an example of wgpu https://github.com/tauri-apps/wry/blob/4f42acb4bd5ea7f16bf0ed18ea8113a9bd957659/examples/wgpu.rs

Can this be done in a regular Tauri app?

**Comment by amrbashir** [comment]
Not yet but hopefully will be in v2 or worst case in v3

**Comment by igagen** [comment]
> Not yet but hopefully will be in v2 or worst case in v3

Great! This will be an amazing capability that will really set Tauri apart from the competition.

**Comment by jtoy** [comment]
Amazing work, thanks for this!

**Comment by ewmb7701** [comment]
> Not yet but hopefully will be in v2 or worst case in v3

@amrbashir Is this landing in v2, or should I use wry directly? I want to use Tauri for cross-platform game development.

**Comment by Manubi** [comment]
> > Not yet but hopefully will be in v2 or worst case in v3
> 
> @amrbashir Is this landing in v2, or should I use wry directly? I want to use Tauri for cross-platform game development.

Would also be interested in this. Is there a simple example somewhere?

**Comment by lukigarazus** [comment]
@amrbashir did this land in v2?

**Comment by amrbashir** [comment]
yeah but it is behind `unstable` feature flag, there is now `Window` type that doesn't have a webview in it, so you can put whatever you want there and you can also add `Webview` type in it as well if you like.

**Comment by df007df** [comment]
Whether there is a demo using the Tauri2.0 app ?
Because it's too late to show the video right now
@amrbashir

**Comment by amrbashir** [comment]
well, it is not exactly a winit subwindow, but a tao subwindow which you can use to render wgpu or something inside it, like https://github.com/FabianLars/tauri-v2-wgpu

**Comment by franzcrs** [comment]
Hello. Would it be possible to display an egui UI to the Tauri Window or WebViewWindow ?

**Comment by tqpgun** [comment]
> Hello. Would it be possible to display an egui UI to the Tauri Window or WebViewWindow ?

https://github.com/tauri-apps/tauri-egui

**Comment by franzcrs** [comment]
> > Hello. Would it be possible to display an egui UI to the Tauri Window or WebViewWindow ?
> 
> https://github.com/tauri-apps/tauri-egui

That integration is quite old, and when I tried running the example, I immediately ran into dependency conflicts. Thanks for the reply anyway!

For anyone else interested in getting this working, I was able to render egui in a child window using this standalone integration:

**[tauri-plugin-egui by clearlysid](https://github.com/clearlysid/tauri-plugin-egui)**

Its author hopes the Tauri Team will consider giving it official support.
